### PR TITLE
uuid custom type casting example

### PR DESCRIPTION
**Pull request by jinjagit**
# Custom type casting

This issue was raised during a recent PR, and it was suggested that it should be moved to a draft PR for further discussion / investigation.

This repo. contains a working example (see: ../schema/custom_types/uuid.ex and use of `:uuid` type in schema.ex).

Please continue the discussion in this PR and/or add links to further information / opinion.

## Record of previous discussions re. custom type casting:

@oliverbarnes [comment](https://github.com/liquidvotingio/api/pull/71#discussion_r452067627):

Interesting, this is new to me. What are the benefits of using a custom type and checking for the correct uuid format at the graphql layer (as opposed to having Ecto barf when the format is bad)?

@niklaslong [comment](https://github.com/liquidvotingio/api/pull/71#discussion_r452083436):

Both would be fine; to me it's about better modelling the data that's passed in/out of the application. Internally, when dealing with a `uuid`, it's of type `Ecto.UUID`. Regardless of whether we're passing in a valid `uuid` or not, it should be consistently represented and that implies casting with `Ecto.UUID.cast/1`. Thoughts?

@oliverbarnes [comment](https://github.com/liquidvotingio/api/pull/71#discussion_r452118967):

Not sure. The issue of representation consistency seems a little abstract to me. Would a string representation for an integer id be cast before letting Ecto handle it?

In practice does Ecto actually require the UUID string to be cast before finding a record with it?

Meant to confirm this myself, but am amidst `mix deps.get` conflicts here at the moment. https://stackoverflow.com/a/46799252/358518 says `Repo.get!` will try to cast the string itself.

@niklaslong [comment](https://github.com/liquidvotingio/api/pull/71#discussion_r452135829):

In that case, we should probably decode it as a `string` and rely on `Ecto` to do the casting thinking
(Here are the absinthe docs on custom scalar types if anyone's curious: https://hexdocs.pm/absinthe/custom-scalars.html)

@oliverbarnes [comment](https://github.com/liquidvotingio/api/pull/71#discussion_r452140870):

Yeah, I did check those out. I'm not against it, I like having a specific type in terms of semantics. I'd just like to understand the practical gain, if any. With the `datetime` custom scalar example it's not clear to me either.

Also good for us to understand how this will affect consumer clients. What's the behavior if the cast fails on our end? I see the `:error` being returned by `decode`, but what's the final response like? 200 OK with error fields, or a 400 Bad Request?

@oliverbarnes [comment](https://github.com/liquidvotingio/api/pull/71#discussion_r453503942):

This isn't resolved :) Two questions to be answered here: what is the gain of having the custom type exactly, and how does it change the final response, if at all?

### CODE REVIEW

**Comment by oliverbarnes** [comment]
Thanks man, this gives a good snapshot of the state of the discussion. Are you in the process of researching this more in depth to answer the questions raised?

File: lib/liquid_voting_web/schema/custom_types/uuid.ex [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    parse(&decode/1)
+  end
+
+  @spec decode(Absinthe.Blueprint.Input.String.t()) :: {:ok, term()} | :error
+  @spec decode(Absinthe.Blueprint.Input.Null.t()) :: {:ok, nil}
+  defp decode(%Absinthe.Blueprint.Input.String{value: value}), do: UUID.cast(value)
+
+  defp decode(%Absinthe.Blueprint.Input.Null{}), do: {:ok, nil}
+
+  defp decode(_), do: :error
```
[the remark this answers is not in the document]
Comment (jinjagit) [reply lib/liquid_voting_web/schema/custom_types/uuid.ex:26]: done.

[the remark this answers is not in the document]
Comment (niklaslong) [reply lib/liquid_voting_web/schema/custom_types/uuid.ex:26]: I'd not space the specs though

File: lib/liquid_voting_web/schema/custom_types/uuid.ex [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    The `UUID` scalar type represents UUID compliant string data, represented as UTF-8
+    character sequences. The UUID type is most often used to represent unique
+    human-readable ID strings.
+    """)
+
+    serialize(&encode/1)
+    parse(&decode/1)
+  end
+
+  @spec decode(Absinthe.Blueprint.Input.String.t()) :: {:ok, term()} | :error
```
Comment (niklaslong) [inline lib/liquid_voting_web/schema/custom_types/uuid.ex:21]:
```suggestion
  @spec decode(Absinthe.Blueprint.Input.String.t()) :: {:ok, term()} | :error}
```

I'm also wondering if we really need the typespecs as these are private functions. I think the functions are clear without, thoughts?

**Comment by niklaslong** [comment]
@oliverbarnes I'm happy to close this for now. If we decide to validated input data in a plug at the GraphQL layer then we should revisit this 👍

**Pull request closed without merging**